pageserver/src/tenant/timeline/delete.rs

```diff
@@ -156,9 +156,9 @@ async fn delete_local_layer_files(
     tenant_id: TenantId,
     timeline: &Timeline,
 ) -> anyhow::Result<()> {
-    info!("waiting for layer_removal_cs.lock()");
-    let layer_removal_guard = timeline.layer_removal_cs.lock().await;
-    info!("got layer_removal_cs.lock(), deleting layer files");
+    info!("waiting for locks");
+    let guards = tokio::join!(timeline.gc_lock.lock(), timeline.compaction_lock.lock());
+    info!("got locks, deleting layer files");
```

Comment:
> does tokio::join! poll the futures in any particular order

no order. I would expect it to lower to select, but did not check.

Reply:
These locks are introduced in this PR and they are guarding against http apis and background task executing the compaction and gc.

I'm adding comments about how these can only be used for these things; I don't think there is an abstraction which would allow guarding against the misuse...

Timeline deletion does stop all tasks before, so it is only guarding against the http api there as far as I can see.

6505dca8b doc: guidance against using per-op locks for other things